realm/realm-library/src/main/cpp/io_realm_internal_OsRealmConfig.cpp

```diff
@@ -349,7 +352,52 @@ JNIEXPORT jstring JNICALL Java_io_realm_internal_OsRealmConfig_nativeCreateAndSe
         config.sync_config->stop_policy = session_stop_policy;
         config.sync_config->error_handler = std::move(error_handler);
         switch (j_client_reset_mode) {
-            case io_realm_internal_OsRealmConfig_CLIENT_RESYNC_MODE_SEAMLESS_LOSS: config.sync_config->client_resync_mode = realm::ClientResyncMode::SeamlessLoss; break;
+            case io_realm_internal_OsRealmConfig_CLIENT_RESYNC_MODE_SEAMLESS_LOSS: {
+                config.sync_config->client_resync_mode = realm::ClientResyncMode::DiscardLocal;
+
+                static JavaClass before_client_reset_handler_class(env, "io/realm/internal/SyncObjectServerFacade$BeforeClientResetHandler");
+                static JavaClass after_client_reset_handler_class(env, "io/realm/internal/SyncObjectServerFacade$AfterClientResetHandler");
+                static JavaMethod on_before_client_reset_method(env, before_client_reset_handler_class, "onBeforeReset","(JJLio/realm/internal/OsRealmConfig;)V", false);
+                static JavaMethod on_after_client_reset_method(env, after_client_reset_handler_class, "onAfterReset","(JLio/realm/internal/OsRealmConfig;)V", false);
+
+                JavaGlobalWeakRef j_config_weak(env, j_config);
+                JavaGlobalWeakRef j_on_before_client_reset_handler_weak(env, j_on_before_client_reset_handler);
+                JavaGlobalWeakRef j_on_after_client_reset_handler_weak(env, j_on_after_client_reset_handler);
+                config.sync_config->notify_before_client_reset = [j_on_before_client_reset_handler_weak, j_config_weak](SharedRealm local_version, SharedRealm remote_version) {
+                    JNIEnv* env = JniUtils::get_env(false);
+                    JavaGlobalRefByMove config_global = j_config_weak.global_ref(env);
+                    if (!config_global) {
+                        return;
+                    }
+
+                    SharedRealm* local_version_ptr = new SharedRealm(local_version);
+                    SharedRealm* remote_version_ptr = new SharedRealm(remote_version);
+                    j_on_before_client_reset_handler_weak.call_with_local_ref(env, [&](JNIEnv* env, jobject obj) {
+                        env->CallVoidMethod(obj, on_before_client_reset_method, reinterpret_cast<jlong>(local_version_ptr), reinterpret_cast<jlong>(remote_version_ptr), config_global.get());
+                        // Fixme before merging, closing the realms should be handled by core?
```

Comment:
I think that this Realm closing should be handled by core, as it is opened there.

These Realm's must be closed because when opened they prevent sessions to be freed, causing the test cases to fail when closing the `App`.

Reply:
Implemented `ManualReleaseNativeContext` that allows to bind the release of its native objects to a scope.